panther_gpiod/src/gpio_driver.cpp

```diff
@@ -0,0 +1,323 @@
+  }
+
+  return settings;
+}
+
+void GPIODriver::ChangePinDirection(const GPIOPin pin, const gpiod::line::direction direction)
+{
+  std::unique_lock lock(gpio_info_storage_mutex_);
+  GPIOInfo & gpio_info = GetGPIOInfoRef(pin);
+
+  if (gpio_info.direction == direction) {
+    return;
+  }
+
+  gpio_info.direction = direction;
+
+  auto line_config = gpiod::line_config();
+
+  for (const auto & gpio_info : gpio_info_storage_) {
+    gpiod::line_settings settings = GenerateLineSettings(gpio_info);
+    line_config.add_line_settings(gpio_info.offset, settings);
+  }
+
+  line_request_->reconfigure_lines(line_config);
+  gpio_info.value = line_request_->get_value(gpio_info.offset);
+}
+
+bool GPIODriver::IsPinActive(const GPIOPin pin)
+{
+  if (!IsGPIOMonitorThreadRunning()) {
+    throw std::runtime_error("GPIO monitor thread is not running!");
+  }
+
+  std::unique_lock lock(gpio_info_storage_mutex_);
+  const GPIOInfo & pin_info = GetGPIOInfoRef(pin);
+  return pin_info.value == gpiod::line::value::ACTIVE;
+}
+
+bool GPIODriver::SetPinValue(const GPIOPin pin, const bool value)
+{
+  GPIOInfo & gpio_info = GetGPIOInfoRef(pin);
+
+  if (gpio_info.direction == gpiod::line::direction::INPUT) {
+    throw std::invalid_argument("Cannot set value for INPUT pin.");
+  }
+
+  gpiod::line::value gpio_value = value ? gpiod::line::value::ACTIVE : gpiod::line::value::INACTIVE;
+
+  std::unique_lock lock(gpio_info_storage_mutex_);
+  try {
+    line_request_->set_value(gpio_info.offset, gpio_value);
+
+    if (line_request_->get_value(gpio_info.offset) != gpio_value) {
+      throw std::runtime_error("Failed to change GPIO state");
+    }
+
+    gpio_info.value = gpio_value;
+    if (GPIOEdgeEventCallback) {
+      GPIOEdgeEventCallback(gpio_info);
+    }
+
+    return true;
+  } catch (const std::exception & err) {
+    std::cerr << "Error while setting GPIO pin value: " << err.what() << std::endl;
+    return false;
+  }
+}
+
+void GPIODriver::GPIOMonitorOn()
+{
+  if (IsGPIOMonitorThreadRunning()) {
+    return;
+  }
+
+  {
+    std::unique_lock lock(gpio_info_storage_mutex_);
+    for (auto & info : gpio_info_storage_) {
+      info.value = line_request_->get_value(info.offset);
+    }
+  }
+
+  std::unique_lock<std::mutex> lck(monitor_init_mtx_);
+
+  gpio_monitor_thread_enabled_ = true;
+  gpio_monitor_thread_ = std::make_unique<std::thread>(&GPIODriver::MonitorAsyncEvents, this);
+
+  if (
+    monitor_init_cond_var_.wait_for(lck, std::chrono::milliseconds(50)) ==
+    std::cv_status::timeout) {
+    throw std::runtime_error("Timeout while waiting for GPIO monitor thread");
+  }
+}
+
+void GPIODriver::MonitorAsyncEvents()
+{
+  if (use_rt_) {
+    ConfigureRt();
+  }
+
+  auto edge_event_buffer = gpiod::edge_event_buffer(edge_event_buffer_size_);
+
+  struct pollfd pollfd;
+  pollfd.fd = line_request_->fd();
+  pollfd.events = POLLIN;
+
+  {
+    std::unique_lock<std::mutex> lck(monitor_init_mtx_);
```

Comment:
```suggestion
    std::lock_guard<std::mutex> lck(monitor_init_mtx_);
```
in this simple case it should be a better fit.